server/src/test/java/org/elasticsearch/index/replication/RecoveryDuringReplicationTests.java

```diff
@@ -363,24 +365,68 @@ public void testResyncAfterPrimaryPromotion() throws Exception {
                 final BulkShardRequest bulkShardRequest = indexOnPrimary(indexRequest, oldPrimary);
                 indexOnReplica(bulkShardRequest, shards, newPrimary);
             }
+
+
+            final int extraDocsToBeTrimmed = randomIntBetween(0, 10);
+            logger.info("--> indexing {} extra docs to be trimmed", extraDocsToBeTrimmed);
+            for (int i = 0; i < extraDocsToBeTrimmed; i++) {
+                final IndexRequest indexRequest = new IndexRequest(index.getName(), "type", "extra_trimmed_" + i)
+                    .source("{}", XContentType.JSON);
+                final BulkShardRequest bulkShardRequest = indexOnPrimary(indexRequest, oldPrimary);
+                // have to replicate to another replica != newPrimary one - the subject to trim
+                indexOnReplica(bulkShardRequest, shards, justReplica);
+            }
+
+            logger.info("--> seqNo primary {} replica {}", oldPrimary.seqNoStats(), newPrimary.seqNoStats());
+
             logger.info("--> resyncing replicas");
             PrimaryReplicaSyncer.ResyncTask task = shards.promoteReplicaToPrimary(newPrimary).get();
             if (syncedGlobalCheckPoint) {
                 assertEquals(extraDocs, task.getResyncedOperations());
             } else {
                 assertThat(task.getResyncedOperations(), greaterThanOrEqualTo(extraDocs));
             }
-            shards.assertAllEqual(initialDocs + extraDocs);
+            List<IndexShard> replicas = shards.getReplicas();
+            for (IndexShard replica : replicas) {
+                logger.info("--> seqNo replica {} {}", replica.nodeName(), replica.seqNoStats());
+            }
+
+            // check all docs on primary are available on replica
+            Set<String> primaryIds = getShardDocUIDs(newPrimary);
+            assertThat(primaryIds.size(), equalTo(initialDocs + extraDocs));
+            for (IndexShard replica : replicas) {
+                Set<String> replicaIds = getShardDocUIDs(replica);
+                Set<String> temp = new HashSet<>(primaryIds);
+                temp.removeAll(replicaIds);
+                assertThat(replica.routingEntry() + " is missing docs", temp, empty());
+                temp = new HashSet<>(replicaIds);
+                temp.removeAll(primaryIds);
+                // yeah, replica has more docs as there is no Lucene roll back on it
+                assertThat(replica.routingEntry() + " has to have extra docs", temp,
+                    extraDocsToBeTrimmed > 0 ? not(empty()) : empty());
+            }
+
+            // check translog on replica is trimmed
+            int translogOperations = 0;
+            try(Translog.Snapshot snapshot = getTranslog(justReplica).newSnapshot()) {
+                Translog.Operation next;
+                while ((next = snapshot.next()) != null) {
+                    translogOperations++;
+                    assertTrue("unexpected op: " + next, next.seqNo() < initialDocs + extraDocs);
```

Comment:
can we strength this and look at doc ids also confirm they have the right seq# + term as on the primary ? (now you can have a seq# that happens to be fine but with the wrong doc)